### PR TITLE
Standardize retriever span outputs to a list of MLflow `Document`s

### PR DESCRIPTION
**Pull request by daniellok-db**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

This PR adds an MLflow `Document` class that is intended to be used to standardize retriever outputs for the UI and other downstream tasks. This is necessary because the UI needs to know where to find the documents in order to render them nicely in the UI.

The `Document` class is essentially the same as LangChain's `Document` (may be confusing, I'm open to other naming suggestions), because several downstream applications already depend on the `page_content` and `metadata["doc_uri" / "chunk_id"]` fields being available*. 

We shouldn't force users to conform to the spec as that might break some existing use-cases, but we should at least document that by following the spec, the UI and eval APIs will work properly.

\* _note on `doc_uri` and `chunk_id` metadata fields: it's actually a little complicated and these field names can be something else depending on what users pass to [mlflow.set_retriever_schema](https://github.com/mlflow/mlflow/blob/82e944620080bb48a0e944b17966f6abf193a1ee/mlflow/models/dependencies_schemas.py#L59-L61). however, i simplified it for the docs because it's more for users who are customizing things on their own. if the user is using `set_retriever_schema`, then we will save the correct metadata fields and things should generally work (cc @sunishsheth2009 correct me if i'm wrong)._

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Screenshot 2024-09-26 at 6 41 07 PM

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added the `mlflow.entities.Document` class to represent retrieved documents within `RETRIEVER` spans. If the span output is a list of these `Document`s, a richer UI visualization will be displayed.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/entities/document.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    An entity used in MLflow Tracing to represent retrieved documents in a RETRIEVER span.
+
+    Args:
+        page_content: The content of the document.
+        metadata: A dictionary of metadata associated with the document.
+        id: The ID of the document.
+    """
+
+    page_content: str
+    metadata: Dict[str, any] = field(default_factory=dict)
```
Comment (harupy) [inline mlflow/entities/document.py:18]:
```suggestion
    metadata: Dict[str, Any] = field(default_factory=dict)
```

File: mlflow/entities/document.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    id: Optional[str] = None
+
+    @classmethod
+    def from_langchain_document(cls, document):
+        # older versions of langchain do not have the id attribute
+        id = document.id if hasattr(document, "id") else None
```
Comment (harupy) [inline mlflow/entities/document.py:24]:
```suggestion
        id = getattr(document, "id", None)
```

File: docs/source/llms/tracing/tracing-schema.rst [hunk 3]
```diff
@@ -277,3 +277,52 @@ The sections below provide a detailed view of the structure of a span.
...
+      -
+    
+    * - **Output**
+      - The output must be of type ``List[`` :py:class:`mlflow.entities.Document` ``]``, or a dict matching the structure of the dataclass\*. 
+        The dataclass contains the following properties:
+
+        * **id** (``Optional[str]``) - An optional unique identifier for the document.
+        * **page_content** (``str``) - The text content of the document.
+        * **metadata** (``Optional[Dict[str,any]]``) - The metadata associated with the document. There are two important metadata keys that are used by the UI and evaluation metrics: 
+
```
Comment (B-Step62) [inline docs/source/llms/tracing/tracing-schema.rst:314]:
```suggestion
        * **metadata** (``Optional[Dict[str,any]]``) - The metadata associated with the document. There are two important metadata keys that are reserved for the MLflow UI and evaluation metrics: 
```

**Review by B-Step62 — APPROVED** [review]
LGTM! I like this minimum abstraction approach🙂

File: mlflow/entities/document.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    @classmethod
+    def from_llama_index_node_with_score(cls, node_with_score):
+        metadata = {}
+        metadata["score"] = node_with_score.get_score()
+        # update after setting score so that it can be
+        # overridden if the user wishes to do so
+        metadata.update(deepcopy(node_with_score.metadata))
+
+        return Document(
```
Comment (harupy) [inline mlflow/entities/document.py:40]:
```suggestion
        return cls(
```

File: mlflow/entities/document.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            id=id,
+        )
+
+    @classmethod
+    def from_llama_index_node_with_score(cls, node_with_score):
+        metadata = {}
+        metadata["score"] = node_with_score.get_score()
+        # update after setting score so that it can be
+        # overridden if the user wishes to do so
+        metadata.update(deepcopy(node_with_score.metadata))
```
Comment (harupy) [inline mlflow/entities/document.py:38]: we can simplify this to:

```suggestion
        metadata = {
            "score": node_with_score.get_score(),
            **deepcopy(node_with_score.metadata),
        }
```

File: docs/source/conf.py [hunk 6]
```diff
@@ -406,6 +406,7 @@
     ("py:class", "array"),
     ("py:class", "boolean"),
     ("py:class", "null"),
+    ("py:class", "any"),
```
Comment (harupy) [inline docs/source/conf.py:409]:
```suggestion
```

**Review by harupy — APPROVED** [review]
Just minor comments, LGTM!

File: docs/source/llms/tracing/tracing-schema.rst [hunk 7]
```diff
@@ -277,3 +277,52 @@ The sections below provide a detailed view of the structure of a span.
...
+      - There are no restrictions on the span inputs
+      -
+    
+    * - **Output**
+      - The output must be of type ``List[`` :py:class:`mlflow.entities.Document` ``]``, or a dict matching the structure of the dataclass\*. 
+        The dataclass contains the following properties:
+
+        * **id** (``Optional[str]``) - An optional unique identifier for the document.
+        * **page_content** (``str``) - The text content of the document.
+        * **metadata** (``Optional[Dict[str,any]]``) - The metadata associated with the document. There are two important metadata keys that are reserved for the MLflow and evaluation metrics:
```
Comment (daniellok-db) [inline docs/source/llms/tracing/tracing-schema.rst:313]:
```suggestion
        * **metadata** (``Optional[Dict[str,any]]``) - The metadata associated with the document. There are two important metadata keys that are reserved for the MLflow UI and evaluation metrics: 
```

**Pull request merged**